subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/ResolutionScopeCapabilitiesHandler.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.internal.artifacts.dsl.dependencies;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.LinkedHashMultimap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import org.gradle.api.Action;
+import org.gradle.api.artifacts.ModuleIdentifier;
+import org.gradle.api.artifacts.dsl.CapabilityHandler;
+import org.gradle.api.internal.artifacts.ImmutableModuleIdentifierFactory;
+import org.gradle.api.internal.notations.ModuleIdentifierNotationConverter;
+import org.gradle.internal.component.external.model.CapabilityDescriptor;
+import org.gradle.internal.typeconversion.NotationParser;
+import org.gradle.internal.typeconversion.NotationParserBuilder;
+
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * A capabilities handler that will merge capabilities defined in the build with those
+ * discovered during dependency resolution. We cannot use the same container because capabilities
+ * discovered in one graph shouldn't leak into those discovered in a different graph, even if
+ * in the same build.
+ */
+public class ResolutionScopeCapabilitiesHandler implements CapabilitiesHandlerInternal {
```

Comment:
So if I understand correctly, we currently support build scope declared capabilities - that is capabilities defined in _this_ project and configuration scoped capabilities - that is capabilities defined through dependencies. But is there a way to pick a different `prefer` between `implementation` and `testImplementation` for example?
Or potentially better, a different prefer between different test configurations which would allow to run the same tests with different classpath?
If that's possible, a sample / test would be good. If it is not, maybe we should consider it?

Reply:
That's not exactly that. The "build scope" handler is the one corresponding to the capabilities declared in the consumer:

```
dependencies {
    capabilities {
        // here you say things about the capabilities you declare, or capabilities of other modules
    }
}
```

This handler shouldn't be mutated after configuration. It corresponds to what the consumer declares, and you don't want the discovered capabilities, those that will be found in transitive dependencies, to be overwriting what you find here. So there's effectively a "mixin" capability handler, which is only for the scope of the resolution of a single configuration, which will merge the capabilities declared in the script with those discovered as we visit the graph.

That's why there's a test that shows that if 2 configurations are used, with different dependencies, and those dependencies declare different preferences, the preferences from one configuration shouldn't leak into the preferences for the second configuration resolved.